coap/api/transport.go

```diff
@@ -0,0 +1,201 @@
+			res.Code = gocoap.Changed
+		}
+		return res
+	}
+}
+
+func observe(svc coap.Service) func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+	return func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+		var res *gocoap.Message
+		if msg.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: msg.MessageID,
+				Token:     msg.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
+
+		cid := mux.Var(msg, "id")
+		publisher, err := coap.Authorize(msg, res, cid)
+
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+			return res
+		}
+
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 1 {
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			err := svc.Unsubscribe(id)
+			if err != nil {
+				res.Code = gocoap.InternalServerError
+			}
+		}
+
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 0 {
+			ch := make(chan mainflux.RawMessage)
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			if err := svc.Subscribe(cid, id, ch); err != nil {
+				res.Code = gocoap.InternalServerError
+				return res
+			}
+			go handleSub(svc, id, conn, addr, msg, ch)
+			res.AddOption(gocoap.Observe, 0)
+		}
+		return res
+	}
+}
+
+func notify(svc coap.Service, id string, conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message, counter *uint32, count *[]byte) error {
+	err := sendMessage(conn, addr, msg, counter, count)
+	if err != nil {
+		return svc.Unsubscribe(id)
+	}
+	return nil
+}
+
+func sendMessage(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message, counter *uint32, count *[]byte) error {
+	*counter++
+	binary.LittleEndian.PutUint32(*count, *counter)
+	msg.SetOption(gocoap.Observe, (*count)[:3])
+	var err error
+	for i := 0; i < 3; i++ {
+		err = gocoap.Transmit(conn, addr, *msg)
+		if err != nil {
+			time.Sleep(5 * time.Millisecond)
+			continue
+		}
+		return nil
+	}
+	return err
+}
+
+func handleSub(svc coap.Service, id string, conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message, ch chan mainflux.RawMessage) {
+	var counter uint32
+	count := make([]byte, 4)
+	ticker := time.NewTicker(24 * time.Hour)
+	res := &gocoap.Message{
+		Type:      gocoap.NonConfirmable,
+		Code:      gocoap.Content,
+		MessageID: msg.MessageID,
+		Token:     msg.Token,
+		Payload:   []byte{},
+	}
+	res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+	res.SetOption(gocoap.LocationPath, msg.Path())
+
+	func() {
+		for {
+			select {
+			case <-ticker.C:
+				res.Type = gocoap.Confirmable
+				if err := notify(svc, id, conn, addr, res, &counter, &count); err != nil {
+					return
+				}
+			case rawMsg, ok := <-ch:
+				if !ok {
+					return
+				}
+				res.Type = gocoap.NonConfirmable
+				res.Payload = rawMsg.Payload
+				if err := notify(svc, id, conn, addr, res, &counter, &count); err != nil {
+					return
+				}
+			}
+		}
+	}()
```

Comment:
Why do you need anonymous function here? Maybe you are missing `go` in front?

Reply:
I call it as goroutine elsewhere in the code. I've used an anonymous function in order to avoid labelled loop; I used return instead of `break`, since `break` would only break corresponding `select`.

Reply:
That is ok, but it seems to me that this should be separate (named) function in that case. IMHO it will improve code readability.